### PR TITLE
chore: upgrade Smithy to 1.22

### LINKED ISSUE #599: Upgrade to Smithy 1.18

### PR DESCRIPTION
**Pull request by ianbotsf**
## Issue \#

Closes #599 

## Description of changes

This change upgrades from Smithy 1.17 to 1.22:
* The biggest change between the versions is the deprecation of `set` shapes and their replacement with `list` shapes annotated with the `@uniqueItems` trait. Accordingly, this PR updates several uses of set-based classes from Smithy and adds an automatic translator to replace legacy `set` shapes in model files with new lists.
* Additionally, `CodeWriter` is deprecated and replaced with `AbstractCodeWriter` and a default `SimpleCodeWriter` implementation.

**Companion PR**: [aws-sdk-kotlin#647](https://github.com/awslabs/aws-sdk-kotlin/pull/647)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by lucix-aws — APPROVED** [review]

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 1]
```diff
@@ -71,11 +71,20 @@ class KotlinJmespathExpressionVisitor(val writer: KotlinWriter) : ExpressionVisi
     }
 
     override fun visitAnd(expression: AndExpression): String {
-        throw CodegenException("AndExpression is unsupported")
+        writer.addImport(RuntimeTypes.Utils.truthiness)
```
Comment (lucix-aws) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:74]: No `#T` formatter in the writer?

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 2]
```diff
@@ -71,11 +71,20 @@ class KotlinJmespathExpressionVisitor(val writer: KotlinWriter) : ExpressionVisi
...
+
+        val left = expression.left
```
Comment (lucix-aws) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:76]: It's a nit for sure, but these `left`/`right` temporaries don't seem warranted.

**Comment by ianbotsf** [comment]
> IMO it's worth considering actually generating full boolean expressions instead of generating a series of temporary variables that cascade into evaluating the value of the expression. This would allow us to take advantage of short-circuiting as these waiter expressions could get more complex.

Interesting idea. So right now we generate something like this for OR:

```kotlin
val foo = it?.foo
val fooTruthiness = truthiness(foo)
val bar = it?.bar
val or = if (fooTruthiness) foo else bar
```

Are you suggesting a more compact version of this that omits some/all of `foo`/`fooTruthiness`/`bar`? Something like:

```kotlin
val or = if (truthiness(it?.foo)) it?.foo else it?.bar
```

If so, it'd potentially short-circuit an access of `it?.bar` (the right-hand operand, basically). Is that the general idea?

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 2 above)
Comment (ianbotsf) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:76]: You're right, these were originally more complicated and can be omitted now.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 1 above)
Comment (ianbotsf) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:74]: No, the existing `addTempVar` method accepts a literal code block and doesn't accept/pass placeholder args down to the writer. I could expand it but we've already pushed the scope of this PR far more than I'd like. I'd prefer to keep it simple for now.

**Comment by lucix-aws** [comment]
> > IMO it's worth considering actually generating full boolean expressions instead of generating a series of temporary variables that cascade into evaluating the value of the expression. This would allow us to take advantage of short-circuiting as these waiter expressions could get more complex.
> 
> Interesting idea. So right now we generate something like this for OR:
> 
> ```kotlin
> val foo = it?.foo
> val fooTruthiness = truthiness(foo)
> val bar = it?.bar
> val or = if (fooTruthiness) foo else bar
> ```
> 
> Are you suggesting a more compact version of this that omits some/all of `foo`/`fooTruthiness`/`bar`? Something like:
> 
> ```kotlin
> val or = if (truthiness(it?.foo)) it?.foo else it?.bar
> ```
> 
> If so, it'd potentially short-circuit an access of `it?.bar` (the right-hand operand, basically). Is that the general idea?

Correct.

**Comment by ianbotsf** [comment]
> > > IMO it's worth considering actually generating full boolean expressions instead of generating a series of temporary variables that cascade into evaluating the value of the expression. This would allow us to take advantage of short-circuiting as these waiter expressions could get more complex.
> > 
> > Interesting idea. So right now we generate something like this for OR:
> > ```kotlin
> > val foo = it?.foo
> > val fooTruthiness = truthiness(foo)
> > val bar = it?.bar
> > val or = if (fooTruthiness) foo else bar
> > ```
> > 
> > Are you suggesting a more compact version of this that omits some/all of `foo`/`fooTruthiness`/`bar`? Something like:
> > ```kotlin
> > val or = if (truthiness(it?.foo)) it?.foo else it?.bar
> > ```
> > 
> > If so, it'd potentially short-circuit an access of `it?.bar` (the right-hand operand, basically). Is that the general idea?
> 
> Correct.

We discussed skipping intermediate variables altogether when initially implementing waiters. It would definitely have the benefit of making code smaller, less verbose, and possibly easier to read under simple path specifications (which account for a large portion of the waiters).

We ultimately decided to use granular intermediate variables for the following reasons:
* Without them, the code would become quite lengthy and difficult to mentally parse for complex path expressions. (Some waiters have dozens of subexpressions, all of which could be crowded onto a single line. This would be pretty painful to read much less debug.)
* In-process performance isn't meaningfully impacted by the more verbose code. (Most of these temporary variables will be optimized away by the compiler anyway. More importantly, we've just completed a network IO wait and deserialization run which dwarf the time needed to perform these in-process calculations.)
* Keeping each intermediate calculation in a separate variable makes it easier to trace-log the calculations for aid in debugging. (Note that the trace logging itself isn't added—this is merely a future possibility.)

I'm inclined to apply the same logic to these new `if`/`else` expressions for AND and OR. Short-circuiting is a new benefit but doesn't outweigh the downsides to me. In waiter evaluations, all the data we'd be passing through `if`/`else` blocks is already in memory. Accessing it is fast, cheap, and free of side effects. Enhanced readability/debuggability and (future) granular trace-logging capabilities are far better gains in this situation.

**Pull request merged**